### ISSUE TITLE
Adafactor optimizer slow (~10 s/iteration) compared to SGD

### ISSUE THREAD

**Issue by bkkaggle** [pytorch/xla#1777]
## 🐛 Bug

Adafactor on pytorch/xla is slow, it takes ~5 min for the first few iterations then ~10s/iteration for a full forward+backwards pass compared to SGD, which can do ~10 iterations/second.

## To Reproduce

Here's the colab notebook to reproduce this and create the debugging report:

https://colab.research.google.com/drive/1DF0pnIjsKiq9drUTskeg2dZL8jnFjolO

## Expected behavior

The adafactor optimizer shouldn't be slower than something like SGD, 

## Environment

 - reproducible on XLA backend [CPU/TPU]: TPU
 - torch_xla version: nightly
 - Any other relevant information. In case of issues related to slow executions, re-compilations, etc.., it is useful to have the `tar.gz` report from https://github.com/pytorch/xla/blob/master/TROUBLESHOOTING.md#using-debug_runpy-to-collect-debug-information

## Additional context

Issue was first opened [here](https://github.com/PyTorchLightning/pytorch-lightning/issues/1098#issuecomment-598966152)

**Comment by dlibenzi** [comment]
This is your issue:
```Python
    def _rms(self, x):
        return sqrt(torch.mean(x.pow(2)))
```

You are calling Python's `sqrt()` which triggers an exit from graph-mode into python-scalar-mode.

**Comment by dlibenzi** [comment]
Also, that model is calling `loss.item()` at every step, which is not advised if you are looking at best performance.
That will execute the forward twice (once alone, at `loss.item()`) and once fused with the backward.
We suggest reporting loss values every N steps:

https://github.com/pytorch/xla/blob/890f18c4fc20a0f80437cb6532d01c4936eed307/test/test_train_mp_mnist.py#L129

**Comment by bkkaggle** [comment]
I updated the code to replace the `sqrt`, `copy`, and `functools` calls with pytorch equivalents

```python
    def _rms(self, x):
        return torch.sqrt(torch.mean(x.pow(2)))
```

```python
            # new_shape = (shape[0] * functools.reduce(operator.mul, temp_shape[tmp_div:], 1),
            #              shape[1] * functools.reduce(operator.mul, temp_shape[:tmp_div], 1))
            new_shape = (shape[0] * torch.prod(temp_shape[tmp_div:]),
                         shape[1] * torch.prod(temp_shape[:tmp_div]))
        # return new_shape, copy(shape)
        return new_shape, shape.clone()

```

but the model's speed is still about 5 seconds/iteration. 

In the official Tensorflow adafactor implementation (https://github.com/tensorflow/tensor2tensor/blob/master/tensor2tensor/utils/adafactor.py#L227) it looks like XLA is known to have problems with fusing computations, which may also be happening here.

**Comment by dlibenzi** [comment]
Leave the shapes alone. That is fine.
Change this:
```Python
u.div_(torch.max(self._rms(u) / group['cliping_threshold'], torch.tensor(1.0, device=u.device)))
```

**Comment by dlibenzi** [comment]
I am getting 2.5it/s with that.
And still we are doing extra item() per step.
